### PR TITLE
Refactor code server to use a single queue

### LINKED ISSUE #7466: Loading a module  that contains on_load may hang
**Describe the bug**
If module `foo` contains an `on_load` annotation, then a call to `code:load_binary/3`  (and probably `code:load_abs/1,2`) may hang if there are concurrent calls to these functions, or even `code:ensure_loaded/1`) on the same module.

**To Reproduce**
Given these two modules:

```[erlang]
%% file: foo.erl
-module(foo).

-on_load(init_module/0).

-export([bar/0]).

bar() -> ok.

init_module() ->
    io:format("Initializing ~p....~n", [?MODULE]),
    % wait some time, to make the race condition likely to happen
    timer:sleep(2_000),
    io:format(" done initializing ~p!~n", [?MODULE]).
```

```[erlang]
%% file: repro.erl
-module(repro).

-export([go/0, do_load_binary/1]).

do_load_binary(Parent) ->
  {ok, Bin} = file:read_file("foo.beam"),
  code:load_binary(foo, "foo.beam", Bin),
  Parent ! {done, self()}.

go() ->
  Self = self(),
  X = spawn(fun() -> do_load_binary(Self) end),
  Y = spawn(fun() -> do_load_binary(Self) end),
  A = receive
    {done, X} -> ok
    after 10_000 -> not_ok
  end,
  B = receive
    {done, Y} -> ok
    after 10_000 -> not_ok
  end,
  {A, B}.
```

Run :
```
$ erlc foo.erl repro.erl && erl -noshell -eval 'Res = repro:go(), io:format("~p~n", [Res]), init:stop()'
```

**Expected behavior**
Both concurrent calls to `code:load_binary/3` in `repro:go/0` should eventually return, so we expect `go()` to return `{ok, ok}`. Instead, we see:

```
Initializing foo....
 done initializing foo!
Initializing foo....
 done initializing foo!
{ok,not_ok}
```

Without the `after` clause, the second receive will wait forever.

**Affected versions**
Both R25 and R26 show this problem, I didn't try it on older versions.

**Additional context**
As far as I can see, the problem is in `code_server:try_finish_module`, since the action that is passed to `handle_pending_on_load` will always return a `{noreply, ...}` when the module to load contains an `on_load`; however, if the action ends up waiting since there is an on_load in progress, this means that no reply will ever be sent to the caller.

We found the race condition investigating tests that would sometimes block. In our case, the race was between `code:ensure_loaded/1` and `code:load_binary/3` and the issue happens, again, when the action for `load_binary` waits on the on_load from `ensure_loaded` to finish, but this combination is harder to reproduce consistently.

### LINKED ISSUE #8510: Crash in code_server.erl (badmatch: error)
**Describe the bug**
In certain scenarios (Our massive Purescript codebase compiling into Erlang), possibly pertaining to short lived processes coupled with modules with larger than normal load times but I'm not sure, we get a bad match thrown inside code_server. The place where this reproduces most is in our Cowboy use, if we make some small requests just after start-up, we get the error with the module ```cowboy_static@ps``` (which nearly everything references). The file itself is nothing remarkable and I don't think is related to the issue.

If I preload the module with a manual load this does not happen.

I think this relates to the changes made in #7503 and #6736 

**To Reproduce**
I can't easily build an isolated example

**Expected behavior**
The code should load and not crash code_server

**Affected versions**
26+

**Additional context**
This bit I can help with, I've pulled down latest and built, the problem exists on [line 1184](https://github.com/erlang/otp/blob/79bc8234396a4f619f40dadfb8458064ab29aa85/lib/kernel/src/code_server.erl#L1184) of code_server.erl, we do a ```maps:take``` on the loading collection and the module doesn't exist in there.

A trace of the messages to and from the code server pid reveals the following order of events

```
4672: <9432.50.0> send: {code_server,{{binary_data,3620},
              "/private/var/tmp/_bazel_robashton/01ea9ac11aff3c301299cf1e989e0d76/execroot/__main__/bazel-out/darwin_arm64-fastbuild/bin/media-shell.runfiles/__main__/media-shell_deps/norsk/e
bin/cowboy_static@ps.beam",
              #Ref<0.167300051.2764308486.137148>}} -> <0.1045.0>
4672: <9432.50.0> 'receive': {code_call,<0.1045.0>,
           {load_module,#Ref<0.167300051.2764439556.148827>,cowboy_static@ps,
                        "/private/var/tmp/_bazel_robashton/01ea9ac11aff3c301299cf1e989e0d76/execroot/__main__/bazel-out/darwin_arm64-fastbuild/bin/media-shell.runfiles/__main__/media-shell_de
ps/norsk/ebin/cowboy_static@ps.beam",
                        false,#Ref<0.167300051.2764308486.137148>}}
4673: <9432.50.0> 'receive': {{'LOADER_DOWN',{cowboy_static@ps,{binary_data,3620},
                                  "/private/var/tmp/_bazel_robashton/01ea9ac11aff3c301299cf1e989e0d76/execroot/__main__/bazel-out/darwin_arm64-fastbuild/bin/media-shell.runfiles/__main__/medi
a-shell_deps/norsk/ebin/cowboy_static@ps.beam"}},
 #Ref<0.167300051.2764308486.137148>,process,<0.1045.0>,shutdown}
4673: <9432.50.0> spawn: <0.1047.0>, {erlang,apply,
                    [{"fun code_server:-handle_on_load/5-fun-0-/0",
                      [cowboy_static@ps]},
                     []]}
4673: <9432.50.0> send_to_non_existing_process: {code_server,{{binary_data,3620},
              "/private/var/tmp/_bazel_robashton/01ea9ac11aff3c301299cf1e989e0d76/execroot/__main__/bazel-out/darwin_arm64-fastbuild/bin/media-shell.runfiles/__main__/media-shell_deps/norsk/e
bin/cowboy_static@ps.beam",
              #Ref<0.167300051.2764308481.176919>}}, <0.1046.0>
4673: <9432.50.0> 'receive': {{'LOADER_DOWN',{cowboy_static@ps,{binary_data,3620},
                                  "/private/var/tmp/_bazel_robashton/01ea9ac11aff3c301299cf1e989e0d76/execroot/__main__/bazel-out/darwin_arm64-fastbuild/bin/media-shell.runfiles/__main__/medi
a-shell_deps/norsk/ebin/cowboy_static@ps.beam"}},
 #Ref<0.167300051.2764308481.176919>,process,<0.1046.0>,noproc}
4673: <9432.50.0> 'receive': {'DOWN',#Ref<0.167300051.2764308481.176918>,process,<0.1047.0>,ok}
4673: <9432.50.0> send: {finish_after_on_load,{cowboy_static@ps,true},
                      <0.50.0>,#Ref<0.167300051.2764308481.176922>} -> erts_code_purger
4675: <9432.50.0> 'receive': {reply,finish_after_on_load,ok,#Ref<0.167300051.2764308481.176922>}
4675: <9432.50.0> 'receive': timeout
```

Essentially we get some LOADER_DOWN message, remove the module from the map, and then try to read the module from the map.

Sticking a case statement around the maps:take "fixes" this, but I don't know if that is what is needed, or if it's just a band-aid - not fully understanding the end to end flow of this code and not having sat down with pen and paper to work it out yet. If this is obvious to somebody else then I'm happy for then to make the fix, or tell me what the fix is and I'll do it.

Cheers,

Rob

### PR DESCRIPTION
**Pull request by josevalim**
Prior to this patch, the code server had two
internal queues, one to track module loading
and another to track on_load callbacks. This
pull requests refactors the code to have a
single queue, in order to fix bugs and improve
maintainability.

Closes #7466.
Closes #8510.

### CODE REVIEW

**Comment by josevalim** [comment]
Ideally this would be merged on `maint` because it fixes a regression on OTP 27 but it is a conceptually large change. My suggestion is to merge it to master and, once 27.1 is out, backport it to maint, which should give us time to battle test it for 27.2.

File: lib/kernel/test/code_SUITE.erl
Comment (josevalim) [inline lib/kernel/test/code_SUITE.erl:1456]: This does not change the behaviour of the test. It is simply to make this test share more lines with the test below. :)

File: lib/kernel/src/code_server.erl
Comment (michalmuskala) [inline lib/kernel/src/code_server.erl:1314]: I'm not sure this logic here is optimal. If I understand correctly, with this implementation if we have multiple processes trying to load the same module we'll run `get_object_code` for the first one, and then queue the others. If we fail to resolve the module, we'll reply to the first process with an error, but then we'll still call `get_object_code` for every queued process again and again. Is this intentional? Wouldn't it make sense to reply with an error to all of them once we saw the issue once?

It's somewhat less of a problem for the success case since the `module_loaded` check is very cheap, but the `get_object_call` call is quite expensive.

Comment (josevalim) [reply lib/kernel/src/code_server.erl:1314]: Unfortunately, if we don't do this, we have race conditions. Imagine that several processes attempt to load the same module, and then, one at the very end, do this:

```
file:write_file("app/some_module.beam", ContentsOfBEAM),
code:ensure_loaded(some_module).
```

The `code:ensure_loaded` call should return `{ok, _}`, but with the changes introduced in OTP 26, it may return false.

**Comment by kikofernandez** [comment]
Thanks for your contribution. I plan to look at this after ICFP (in 10 days) approx

**Comment by josevalim** [comment]
False alarm, I had mixed my Erlang versions. Overall, I didn't measure any performance degradation in my usual code loading benchmarks. :)

**Comment by kikofernandez** [comment]
Looking into it this week

**Comment by kikofernandez** [comment]
This was on hold due to other investigations regarding current issues in code_server.
As soon as I am given green light, I review this PR

File: lib/kernel/src/code_server.erl
Comment (lucioleKi) [inline lib/kernel/src/code_server.erl:1358]: Dialyzer seems to complain here: `code_server.erl:1358:36: Expression produces a value of type {'code_server', _}, but this value is unmatched`. Could you fix it (or find a way to dodge it) so that we can put this branch into testing?

Comment (kikofernandez) [reply lib/kernel/src/code_server.erl]: fixed! I will place again for testing

**Review by kikofernandez — APPROVED** [review]
Looks good to me
These changes make the `code_server` easier to follow.
We are running some internal tests and will merge next week if all is green

**Comment by josevalim** [comment]
Glad to hear so! 🎉 Code loading is a state machine and the previous version was encoding the states via anonymous functions and in two separate queues, which made it very hard to follow. :)

**Comment by kikofernandez** [comment]
Completely agree, it was pretty tricky to follow the logic. This was a great improvement.

Since the next OTP release 27.2 is around beginning of December, I am wondering if we can place this already on `maint`, instead of `master`. That gives us close to two months of testing

**Comment by josevalim** [comment]
As long as we are comfortable with potentially doing a 27.2.1 with this commit reverted, it sounds good to me. :D

**Comment by kikofernandez** [comment]
Update. This will be merged in `maint` after I add some types to `X` in `state { loading = #{} :: #{module() => [{X :: term(), pid()}]}}`.

**Comment by bjorng** [comment]
Please do some squashing of the commits before merging this. At least squash the second commit into the first, or squash all commits into one.

**Comment by josevalim** [comment]
@kikofernandez let me know if you want me to do the squashing. :)

**Comment by kikofernandez** [comment]
sure, go ahead! I will merge if our internal builds look good after lunch

**Pull request merged**
